### ISSUE TITLE
Enhancement: [no-misused-promises] / [no-floating-promises] - error when JSON.stringify is called with a promise argument

### ISSUE THREAD

**Issue by elrob** [typescript-eslint/typescript-eslint#7133]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-misused-promises

### Description

`[no-misused-promises]` or `[no-floating-promises]`
I propose that either of these rules should catch mistakes caused by JSON stringifying a promise when intending to JSON stringify the promise's resolved value. This can occur when a method is updated to become async but the caller is not updated to handle it.

### Fail

```typescript
const myMethod = (): Promise<Record<string, string>> => Promise.resolve({my: 'object'});

const myString = JSON.stringify(myMethod());
```

### Pass

```typescript
const myMethod = (): Promise<Record<string, string>> => Promise.resolve({my: 'object'});

const myString = JSON.stringify(await myMethod());
```

**Comment by bradzacher** [comment]
The problem here is that `JSON.stringify` accepts `any` as its argument.

So there's no signal in the types which declare this as a function that shouldn't receive a promise.

There are a number of APIs that exist in a similar style and they face the same problem.

The question is really "should we be in the business of policing what you pass to a function that accepts `any`"? Well that comes down to trying to untangle developer intent. The answer is "in some cases, maybe, in others, no".

In this case we probably could - but I'm not sure if we necessarily should. It's a pretty niche case across the ecosystem - so it's not super high value to lint for it.
I think we'll leave this open to evaluate community engagement.

**Comment by elrob** [comment]
yeah, that's totally fair.
Do you happen to know of any JSON stringify lib/method which has stricter types limited to things that one would be intending to stringify?

**Comment by Josh-Cena** [comment]
I don't think it's possible to define `JSON.stringify` to remove `Promise`, because `Promise` is exactly a subtype of `object` (and I believe you would want to stringify plain objects). Maybe you could use some conditional type tricks, but they don't usually play well with generics and things, and can be easily worked around with "safe-seeming" up-casting.

**Comment by elrob** [comment]
> ts-reset is exactly that! ✨

Thanks @JoshuaKGoldberg . But it seems you're pointing to JSON.parse and there's no JSON.stringify support in there?

**Comment by elrob** [comment]
Thanks for the responses.

I think I'm just going to start wrapping JSON.stringify in typed wrapper functions like this:

```
const serialize = (myObject: MyObjectType): string => JSON.stringify(myObject);
```

Then typing should help when I accidentally pass `Promise<MyObject>` to it.

**Comment by JoshuaKGoldberg** [comment]
> Thanks @JoshuaKGoldberg . But it seems you're pointing to JSON.parse and there's no JSON.stringify support in there?

Ah, good point - might make sense to file an issue on them for using `unknown`. Which, as noted, still won't fix this issue.
